# Implementasi K-NN Algorithm (Tugas 3)


1.   Baca data bunga *iris*
2.   Implementasi algoritma K-NN untuk mencari kelas dari data tes



## Baca Data 



*   Baca data dengan *library pandas* *italicized text*
*   Data [Bunga Iris](https://gist.githubusercontent.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv)


*   Referensi [Hitung Manual](https://docs.google.com/spreadsheets/d/1D0TOWmwVZpA1Wt7e_PrHWAeUlVXRuj2Y/edit?usp=sharing&ouid=104778244268493691619&rtpof=true&sd=true)


#### Persiapan Lingkungan


*   Impor *train_test_split* dari *sklearn*
*   Impor *KNeighborsClassifier* dari *sklearn*
*   Impor *pandas library*



In [32]:
from sklearn.model_selection import train_test_split #split dataset into train and test data

In [33]:
from sklearn.neighbors import KNeighborsClassifier# Create KNN classifier

In [34]:
import pandas as pd

In [35]:
import warnings
warnings.filterwarnings('ignore', category=UserWarning, append=True)

In [36]:
dataset_url = "https://gist.githubusercontent.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv"

In [37]:
df = pd.read_csv(dataset_url)
df.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

In [38]:
#check number of rows and columns in dataset
df.shape

(150, 5)

### Hitung Data

*   Pisahkan kolom "variety" dari *dataframe*
*   Ambil kolom "variety" sebagai target kolom untuk kategori kelas
*   Pisahkan data latih dengan data tes


```
# data latih (nilai data)
X_train 

# data tes (nilai data)
X_test 

# data latih (kelas data)
y_train

# data tes (kelas data)
y_test
```





In [39]:
#create a dataframe with all training data except the target column
X = df.drop(columns=["variety"])#check that the target variable has been removed
X.head()

sepal.length  sepal.width  petal.length  petal.width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [40]:
# separate target values
y = df["variety"].values  # view target values
y[0:5]

array(['Setosa', 'Setosa', 'Setosa', 'Setosa', 'Setosa'], dtype=object)

In [41]:
# percent_amount_of_test_data = / HUNDRED_PERCENT
percent_amount_of_test_data = 0.2

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = percent_amount_of_test_data, random_state=1, stratify=y)



*   Tentukan nilai K = 10 (10 tetangga terdekat)
*   Buat data latih


```
knn_classifier.fit(train_data)
```




In [43]:
amount_of_neighbor = 3

In [44]:
# Fit the classifier to the data
knn = KNeighborsClassifier(n_neighbors = amount_of_neighbor)

# create train data
knn.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=3)



*   Tes data target terhadap data latih
*   Tampilkan predksi 10 data pertama



```
knn_classifier.predict(test_data)
```





In [47]:
#show first 5 model predictions on the test data
knn.predict(X_test)

array(['Virginica', 'Setosa', 'Versicolor', 'Setosa', 'Setosa', 'Setosa',
       'Virginica', 'Virginica', 'Virginica', 'Versicolor', 'Setosa',
       'Versicolor', 'Virginica', 'Versicolor', 'Virginica', 'Setosa',
       'Virginica', 'Versicolor', 'Versicolor', 'Virginica', 'Versicolor',
       'Versicolor', 'Setosa', 'Setosa', 'Virginica', 'Versicolor',
       'Setosa', 'Setosa', 'Versicolor', 'Versicolor'], dtype=object)



*   Cetak prediksi nilai dari data tes



```
knn_classifier.score(test_data)
```





In [46]:
#check accuracy of our model on the test data
knn.score(X_test, y_test)

0.9666666666666667

*  Cek sebuah data dengan nilai 


```
[
    [5.5, 3, 5, 2.2]
]
```



In [55]:
# Custom value to predict
knn.predict([[5.5,	3,	5,	2.2]])

array(['Virginica'], dtype=object)

### Kesimpulan


*   Dataset bunga *iris* terdiri dari 150 data
*   Ambil 80% dari *iris dataset* sebagai data latih
*   Ambil 20% dari *iris dataset* sebagai data tes
*   Tetapkan nilai K = 10
*   Hasil 10 data pertama dari 20% data latih


```
['Virginica', 'Setosa', 'Versicolor', 'Setosa', 'Setosa', 'Setosa',
       'Virginica', 'Virginica', 'Virginica', 'Versicolor']
```
*   Dengan akurasi data yaitu $ 0.967\% $



